apollo-router/src/state_machine.rs

```diff
@@ -227,16 +230,18 @@ impl<FA: RouterSuperServiceFactory> State<FA> {
         match self {
             Running {
                 server_handle: Some(server_handle),
-                mut all_connections_stopped_signal,
+                all_connections_stopped_signals,
                 ..
             } => {
                 tracing::info!("shutting down");
                 let state = server_handle
                     .shutdown()
                     .map_ok_or_else(Errored, |_| Stopped)
                     .await;
-                //FIXME: we might want to set a timeout here
-                let _ = all_connections_stopped_signal.recv().await;
+                for mut receiver in all_connections_stopped_signals.into_iter() {
```

Comment:
maybe a join on all futures would be better?

Reply:
I tried that and it doesn't like the fact that the `mut receiver` doesn't live long enough...

Reply:
I don't think there's a way to join on the futures because you can only create them with a mutable reference to the receiver. e.g.: I tried:
```
                let futs: futures::stream::FuturesUnordered<_> = all_connections_stopped_signals
                    .into_iter()
                    .map(|mut receiver| receiver.recv())
                    .collect();
                let results: Vec<_> = futs.collect().await;
 ```
but that won't compile since:
```
error[E0515]: cannot return reference to function parameter `receiver`
   --> apollo-router/src/state_machine.rs:243:41
    |
243 |                     .map(|mut receiver| receiver.recv())
    |                                         ^^^^^^^^^^^^^^^ returns a reference to data owned by the current function

```
If you have a concrete suggestion, let me know, otherwise I'll go with what I have.

Reply:
@garypen What if you use `iter_mut` instead of `into_iter` ?

Reply:
Ok. I thought a bit about ownership and figured this out.

Reply:
Our thoughts crossed in the ether. That's what I'm doing. :)